File: go/callwire/registry.go
package callwire

import (
	"context"
	"fmt"
	"net"
	"sync"
)

// RegistryServer is a simple built-in service registry.
type RegistryServer struct {
	mu       sync.RWMutex
	services map[string][]string
}

// NewRegistryServer creates a new RegistryServer.
func NewRegistryServer() *RegistryServer {
	return &RegistryServer{
		services: make(map[string][]string),
	}
}

// Register adds a service address.
func (r *RegistryServer) Register(serviceName, addr string) error {
	r.mu.Lock()
	defer r.mu.Unlock()
	addrs := r.services[serviceName]
	for _, a := range addrs {
		if a == addr {
			return nil
		}
	}
	r.services[serviceName] = append(addrs, addr)
	return nil
}

// Discover returns all addresses for a service.
func (r *RegistryServer) Discover(serviceName string) ([]string, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	addrs, ok := r.services[serviceName]
	if !ok || len(addrs) == 0 {
		return nil, fmt.Errorf("service not found: %s", serviceName)
	}
	return addrs, nil
}

// StartRegistryServer starts a callwire server containing the registry endpoints.
func StartRegistryServer(addr string) (net.Listener, error) {
	reg := NewRegistryServer()
	
	Export("callwire.register", reg.Register)
	Export("callwire.discover", reg.Discover)

	l, err := net.Listen("tcp", addr)
	if err != nil {
		return nil, err
	}
	
	go func() {
		for {
			conn, err := l.Accept()
			if err != nil {
				return
			}
			go handleConnection(conn)
		}
	}()
	
	return l, nil
}

// DiscoverPool maintains a dynamic pool of clients resolved from a registry.
type DiscoverPool struct {
	registryAddr string
	serviceName  string
	mu           sync.Mutex
	clients      []*Client
	addrs        []string
}

// NewDiscoverPool creates a new DiscoverPool.
func NewDiscoverPool(registryAddr, serviceName string) (*DiscoverPool, error) {
	pool := &DiscoverPool{
		registryAddr: registryAddr,
		serviceName:  serviceName,
	}
	if err := pool.Refresh(); err != nil {
		return nil, err
	}
	return pool, nil
}

// Refresh queries the registry and updates the active clients.
func (p *DiscoverPool) Refresh() error {
	p.mu.Lock()
	defer p.mu.Unlock()

	regClient, err := Connect(p.registryAddr)
	if err != nil {
		return fmt.Errorf("failed to connect to registry: %w", err)
	}
	defer regClient.Close()

	var addrs []string
	addrs, err = Import[[]string](regClient, context.Background(), "callwire.discover", []interface{}{p.serviceName})
	if err != nil {
		return fmt.Errorf("discover failed: %w", err)
	}

	// Close old clients.
	for _, c := range p.clients {
		c.Close()
	}
	p.clients = nil
	p.addrs = nil

	for _, addr := range addrs {
		c, err := Connect(addr)
		if err == nil {
			p.clients = append(p.clients, c)
			p.addrs = append(p.addrs, addr)
		}
	}

	if len(p.clients) == 0 {
		return fmt.Errorf("no healthy service instances found for %s", p.serviceName)
	}

	return nil
}

// Get returns a client (simple round-robin or first available).
func (p *DiscoverPool) Get() (*Client, error) {
	p.mu.Lock()
	defer p.mu.Unlock()
	if len(p.clients) == 0 {
		return nil, fmt.Errorf("no clients available in pool")
	}
	return p.clients[0], nil
}

// Close closes all connections in the pool.
func (p *DiscoverPool) Close() {
	p.mu.Lock()
	defer p.mu.Unlock()
	for _, c := range p.clients {
		c.Close()
	}
	p.clients = nil
}

// ServeRegistry starts the built-in registry server.
func ServeRegistry(addr string) (net.Listener, error) {
	return StartRegistryServer(addr)
}

// RegisterWith registers a service address with a remote registry server.
func RegisterWith(registryAddr, serviceName, addr string) error {
	client, err := Connect(registryAddr)
	if err != nil {
		return err
	}
	defer client.Close()
	_, err = Import[interface{}](client, context.Background(), "callwire.register", []interface{}{serviceName, addr})
	return err
}

// DiscoverRef returns an RPC helper function pointing to a dynamic discovery client pool.
func DiscoverRef[Resp any](pool *DiscoverPool, funcName string) func(args ...interface{}) (Resp, error) {
	return func(args ...interface{}) (Resp, error) {
		var zero Resp
		client, err := pool.Get()
		if err != nil {
			return zero, err
		}
		cleanArgs := args
		if cleanArgs == nil {
			cleanArgs = []interface{}{}
		}
		return Import[Resp](client, context.Background(), funcName, cleanArgs)
	}
}

File: python/callwire/registry.py
import threading
from typing import List, Dict
from .server import export, serve
from .client import Client

class RegistryServer:
    def __init__(self):
        self.mu = threading.Lock()
        self.services: Dict[str, List[str]] = {}

    def register(self, service_name: str, addr: str) -> None:
        with self.mu:
            addrs = self.services.setdefault(service_name, [])
            if addr not in addrs:
                addrs.append(addr)

    def discover(self, service_name: str) -> List[str]:
        with self.mu:
            addrs = self.services.get(service_name)
            if not addrs:
                raise ValueError(f"service not found: {service_name}")
            return addrs

import socket
from .server import export, _handle_connection

def serve_registry(host: str = "localhost", port: int = 9090) -> None:
    reg = RegistryServer()
    export(reg.register)
    export(reg.discover)

    listener = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    listener.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    listener.bind((host, port))
    listener.listen()

    while True:
        try:
            conn, _ = listener.accept()
            t = threading.Thread(target=_handle_connection, args=(conn,), daemon=True)
            t.start()
        except OSError:
            break

class DiscoverPool:
    def __init__(self, registry_addr: str, service_name: str):
        self.registry_addr = registry_addr
        self.service_name = service_name
        self.mu = threading.Lock()
        self.clients: List[Client] = []
        self.addrs: List[str] = []
        self.refresh()

    def refresh(self) -> None:
        with self.mu:
            host, port_str = self.registry_addr.split(":")
            reg_client = Client()
            reg_client.connect(host, int(port_str))
            try:
                addrs = reg_client.call("discover", [self.service_name])
            finally:
                reg_client.close()

            # Close old clients
            for c in self.clients:
                c.close()
            self.clients.clear()
            self.addrs.clear()

            for addr in addrs:
                try:
                    h, p = addr.split(":")
                    c = Client()
                    c.connect(h, int(p))
                    self.clients.append(c)
                    self.addrs.append(addr)
                except Exception:
                    pass

            if not self.clients:
                raise RuntimeError(f"no healthy service instances found for {self.service_name}")

    def get(self) -> Client:
        with self.mu:
            if not self.clients:
                raise RuntimeError("no clients available in pool")
            return self.clients[0]

    def close(self) -> None:
        with self.mu:
            for c in self.clients:
                c.close()
            self.clients.clear()
